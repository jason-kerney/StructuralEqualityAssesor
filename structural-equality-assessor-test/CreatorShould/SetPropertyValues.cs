using System;
using System.Collections.Generic;
using System.Reflection;
using ApprovalTests;
using ApprovalTests.Reporters;
using StructuralEqualityAssessor.Internal;
using StructuralEqualityAssessor.Test.Examples;
using Xunit;

namespace StructuralEqualityAssessor.Test.CreatorShould
{
    [UseReporter(typeof(DiffReporter))]
    public class SetPropertyValues
    {
        [Fact]
        public void OnMultipleObjects()
        {
            var a = new OtherClass();
            var b = new OtherClass();
            var c = new OtherClass();

            var targets = new[] {a, b};
            var (propertyValues, _) = Creator.GetPropertyAndFieldValues(typeof(OtherClass), null);
            Action<PropertyInfo, object> noOp = (p, v) => { };

            Creator.SetPropertyValues(targets, propertyValues, noOp);

            Assert.Equal(a.ToString(), b.ToString());
            Assert.NotEqual(a.ToString(), c.ToString());
        }

        [Fact]
        public void OnMultipleObjectsAndExecutesDelegateOnEachProperty()
        {
            var a = new OtherClass();
            var properties = new List<string>();

            var targets = new[] {a};
            var (propertyValues, _) = Creator.GetPropertyAndFieldValues(typeof(OtherClass), null);
            Action<PropertyInfo, ValueAccessors> collect = (property, value) => properties.Add($"{property} => {value}");

            Creator.SetPropertyValues(targets, propertyValues, collect);

            Approvals.VerifyAll(properties, "result");
        }

        [Fact]
        public void OnComplexObjects()
        {
            var a = new SomeComplexClass();

            var targets = new[] {a};
            var ctors = new Dictionary<Type, Func<object>>
            {
                {typeof(NonDefaultConstructor), () => new NonDefaultConstructor(6)}
            };
            var (propertyValues, _) = Creator.GetPropertyAndFieldValues(typeof(SomeComplexClass), ctors);

            Creator.SetPropertyValues(targets, propertyValues, (p, v) => { });

            Approvals.Verify(a);
        }

        [Fact]
        public void OnSingleObject()
        {
            var target = new OtherClass();
            var (propertyValues, _) = Creator.GetPropertyAndFieldValues(typeof(OtherClass), null);
            Creator.SetPropertyValues(target, propertyValues);

            Approvals.Verify(target.ToString());
        }
    }
}
