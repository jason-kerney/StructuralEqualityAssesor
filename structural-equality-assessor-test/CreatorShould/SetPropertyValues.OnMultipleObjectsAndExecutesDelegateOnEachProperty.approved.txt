result[0] = Int32 SomeNumber => { Populated: 1, Empty: [0] }
result[1] = System.String SomeString => { Populated: "A string", Empty: [null] }
